azure-iot-device/azure/iot/device/common/pipeline/pipeline_thread.py

```diff
@@ -0,0 +1,199 @@
+  handlers which call into the pipeline to deliver protocol messages.
+
+6. Calls into the callback thread could theretically block, but we currently
+  only have decorators which enter the callback thread without blocking.  This
+  is done to ensure that client code does not execute on the pipeline thread and
+  also to ensure that the pipline thread is not blocked while waiting for client
+  code to execute.
+
+These decorators use concurrent.futures.Future and the ThreadPoolExecutor because:
+
+1. The thread pooling with a pool size of 1 gives us a single thread to run all
+  pipeline operations and a different (single) thread to run all callbacks.  If
+  the code attempts to run a second pipeline operation (or callback) while a
+  different one is running, the ThreadPoolExecutor will queue the code until the
+  first call is completed.
+
+2. The concurent.futures.Future object properly handles both Exception and
+  BaseException errors, re-raising them when the Future.result method is called.
+  threading.Thread.get() was not an option because it doesn't re-raise
+  BaseException errors when Thread.get is called.
+
+3. concurrent.futures is available as a backport to 2.7.
+
+"""
+
+
+def _get_named_executor(thread_name):
+    """
+    Get a ThreadPoolExecutor object with the given name.  If no such executor exists,
+    this function will create on with a single worker and assign it to the provided
+    name.
+    """
+    executor = getattr(_get_named_executor, thread_name, None)
+    if not executor:
+        logger.info("Creating {} executor".format(thread_name))
+        executor = ThreadPoolExecutor(max_workers=1)
+        setattr(_get_named_executor, thread_name, ThreadPoolExecutor(max_workers=1))
+    return executor
+
+
+def _get_thread_local_storage():
+    """
+    Get an object which contains "thread local storage" using the threading.local()
+    function.  If no such object exists, one is created.  The object returned from
+    `threading.local()` appears to be a global object, but has a different value for
+    each individual thread.  We use this so we can check which thread we're running
+    inside at runtime.
+    """
+    if not getattr(_get_thread_local_storage, "local", None):
+        logger.info("Creating thread local storage")
+        _get_thread_local_storage.local = threading.local()
+    return _get_thread_local_storage.local
+
+
+def _invoke_on_executor_thread(thread_name, block=True, _func=None):
+    """
+    Decorator which runs the wrapped function on a given thread.  If block==False,
+    the call returns immediately without waiting for the decorated function to complete.
+    If block==True, the call waits for the decorated function to complete before returning.
+    """
+
+    def decorator(func):
+        function_name = getattr(func, "__name__", str(func))
+
+        def wrapper(*args, **kwargs):
+            if not getattr(_get_thread_local_storage(), "in_{}_thread".format(thread_name), False):
```

Comment:
All these `getattr`s make me nervous, especially when they start mapping to strings to check statuses. `getattr` is one of those language constructs that really isn't supposed to be used except in exceptional circumstances. Could we maybe use some kind of dict for cached values instead?